src/Microsoft.PowerShell.ConsoleHost/host/msh/Executor.cs

```diff
@@ -320,11 +322,346 @@ internal Collection<PSObject> ExecuteCommand(string command, out Exception excep
+                throw ex;
+            }
+        }
+
+        internal class ImplicitRemotingBatchingNotSupportedException : RuntimeException
+        {
+            internal ImplicitRemotingBatchingNotSupportedException(string message) : base(message)
+            { }
+        }
+
+        // Commands allowed to run on target remote session along with implicit remote commands
+        private static readonly HashSet<string> AllowedCommands = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
+        {
+            "ForEach-Object",
+            "Measure-Command",
+            "Measure-Object",
+            "Sort-Object",
+            "Where-Object"
+        };
+
+        // Determines if the typed command invokes implicit remoting module proxy functions in such
+        // a way as to allow simple batching, to reduce round trips between client and server sessions.
+        // Requirements:
+        //  a. All commands must be implicit remoting module proxy commands targeted to the same remote session
+        //  b. Except for *allowed* commands that can be safely run on remote session rather than client session
+        //  c. Commands must be in a simple pipeline
+        private bool TryRunAsImplicitBatch(string command, out Exception exceptionThrown)
+        {
+            exceptionThrown = null;
+
+            // Early out
+            if (_parent.RunspaceRef.IsRunspaceOverridden ||
+                (_parent.RunspaceRef.Runspace.ExecutionContext.Modules != null && !_parent.RunspaceRef.Runspace.ExecutionContext.Modules.IsImplicitRemotingModuleLoaded))
+            {
+                return false;
+            }
+
+            try
+            {
+                var scriptBlock = ScriptBlock.Create(command);
+                var scriptBlockAst = scriptBlock.Ast as ScriptBlockAst;
+                if (scriptBlockAst == null)
+                {
+                    throw new ImplicitRemotingBatchingNotSupportedException(ConsoleHostStrings.ImplicitBatchInvalidScriptBlock);
+                }
+
+                // Make sure that this is a simple pipeline
+                string errorId;
+                string errorMsg;
+                scriptBlockAst.GetSimplePipeline(true, out errorId, out errorMsg);
+                if (errorId != null)
+                {
+                    throw new ImplicitRemotingBatchingNotSupportedException(ConsoleHostStrings.ImplicitBatchScriptNotSimplePipeline);
+                }
+
+                // Run checker
+                var checker = new PipelineForBatchingChecker { ScriptBeingConverted = scriptBlockAst };
+                scriptBlockAst.InternalVisit(checker);
+
+                // If this is just a single command, there is no point in batching it
+                if (checker.Commands.Count < 2)
+                {
+                    throw new ImplicitRemotingBatchingNotSupportedException(ConsoleHostStrings.ImplicitBatchSingleCommand);
```

Comment:
At one point I was thinking of emitting compatibility information through tracing, and do that with exceptions.  But since this is such a rare condition it doesn't seem worth it.  I'll remove them as needed.